packit/local_project.py

```diff
@@ -562,3 +566,486 @@ def fetch(self, remote: str, refspec: Optional[str] = None, force: bool = False)
 
     def __del__(self):
         self.clean()
+
+
+class LocalProjectCalculationState:
+    """Class representing the current state of construction of LocalProject.
+
+    Encloses all the arguments that will then be passed to LocalProject's
+    constructor.
+    """
+
+    def __init__(
+        self,
+        remote: str = "",
+        git_repo: Optional[git.Repo] = None,
+        working_dir: Union[Path, str, None] = None,
+        ref: str = "",
+        git_project: Optional[GitProject] = None,
+        git_service: Optional[GitService] = None,
+        git_url: str = "",
+        full_name: str = "",
+        namespace: str = "",
+        repo_name: str = "",
+        working_dir_temporary: bool = False,
+    ):
+        self.remote = remote
+        self.git_repo = git_repo
+        self.working_dir = working_dir
+        self.ref = ref
+        self.git_project = git_project
+        self.git_service = git_service
+        self.git_url = git_url
+        self.full_name = full_name
+        self.namespace = namespace
+        self.repo_name = repo_name
+        self.working_dir_temporary = working_dir_temporary
+
+    @staticmethod
+    def from_local_project(lp: LocalProject) -> "LocalProjectCalculationState":
+        """Constructs the calculation state from an existing LocalProject."""
+        return LocalProjectCalculationState(
+            lp.remote,
+            lp.git_repo,
+            lp.working_dir,
+            lp.ref,
+            lp.git_project,
+            lp.git_service,
+            lp.git_url,
+            lp.full_name,
+            lp.namespace,
+            lp.repo_name,
+        )
+
+
+# Use an explicit sentinel class to have stricter type-checking than just object()
+class _CalculateType:
+    pass
+
+
+CALCULATE = _CalculateType()
+
+
+T = TypeVar("T")
+CanCalculate = Union[T, _CalculateType]
+
+
+class LocalProjectBuilder:
+    """Class for building instances of LocalProject dynamically."""
+
+    PREREQUISITES = {
```

Comment:
Now I remembered that this hardcoding via strings was something that I didn't really like during my initial implementation but couldn't figure out anything cleaner. Maybe some decorator magic could work but that may be too complicated.